### PR TITLE
bus factor code and testing suite

### PR DESCRIPTION
**Pull request by tianayjlin**
bus factor code and testing suite.
also i had this issue where `npm run jest` threw errors unless i `npm run clean`ed it first. added that line to the package.json  under scripts > jest

### CODE REVIEW

File: src/metrics/BusFactor.ts
Comment (ealoisel) [inline src/metrics/BusFactor.ts:1]: Set value to 0 instead of -1 upon error

File: src/metrics/BusFactor.ts [hunk 1]
```diff
@@ -8,19 +16,97 @@ export class BusFactor extends Metric {
...
+            Logger.logDebug("Bus Factor: Error fetching bus factor data from repository\n" + Error);
+            this.score = -1; 
+            this.endTimer(); 
+            return;
+        }
+
+        // console.log(`\nlargest contributor:${author}\nlines altered: ${hc_lines}/${total_lines}\ncommits:${hc_commits}/${total_commits}`);
+
+        //there are certain github repo settings that omit data
+        if(!total_lines && total_commits){
```
Comment (ealoisel) [inline src/metrics/BusFactor.ts:91]: Does this check for if there are no lines and no commits

File: src/metrics/BusFactor.ts [hunk 2]
```diff
@@ -8,19 +16,97 @@ export class BusFactor extends Metric {
...
+        let hc_lines = 0;
+
+        let author:string = " "
+        try{
+            //make api call to ep to return container of objects.
+            const response = await axios.get(ep, {headers: {'Authorization': `token ${process.env.GITHUB_TOKEN}`}}); 
+            
+            if(!response || !response.data || !response.data.length){
+                console.log('data is null');
+                this.score = -1;
```
Comment (ealoisel) [inline src/metrics/BusFactor.ts]: Set = 0 instead of -1

File: src/test/busFactor.test.ts [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+        
+        //activate spying before you run the program duh
+        const spy = jest.spyOn(console, 'log');
+
+        await bf.calculateScore(); 
+
+        //spy to see if null message is created
+        expect(spy).toHaveBeenLastCalledWith('data is null')
+    
+        expect(bf.getScore()).toBe(-1);
```
Comment (ealoisel) [inline src/test/busFactor.test.ts]: expect to be 0

File: src/test/busFactor.test.ts [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    it('should return -1 if endpoint returns null', async() => {
+        (axios.get as jest.Mock).mockResolvedValueOnce(null); 
+        
+        //activate spying before you run the program duh
+        const spy = jest.spyOn(console, 'log');
+
+        await bf.calculateScore(); 
+
+        //spy to see if null message is created
+        expect(spy).toHaveBeenLastCalledWith('data is null')
```
Comment (ealoisel) [inline src/test/busFactor.test.ts]: Don't print this out to the console or it will break the autograder

File: src/test/busFactor.test.ts [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+        (axios.get as jest.Mock).mockResolvedValueOnce({data:[]}); 
+
+        const spy = jest.spyOn(console, 'log'); 
+
+        //set up mock for axios.get, returns a promise (resolved value once) to an empty arr when called anywhere! obj attributes are included.
+        await bf.calculateScore(); 
+
+        expect(spy).toHaveBeenLastCalledWith('data is null');
+        
+        expect(bf.getScore()).toBe(-1);
```
Comment (ealoisel) [inline src/test/busFactor.test.ts]: 0

File: src/test/busFactor.test.ts [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+              ]
+            }
+          ]};
+
+        (axios.get as jest.Mock).mockResolvedValueOnce(response); 
+        const spy = jest.spyOn(console, 'log'); 
+        
+        await bf.calculateScore(); 
+
+        expect(spy).not.toHaveBeenCalledWith('data is null');
```
Comment (ealoisel) [inline src/test/busFactor.test.ts]: no printouts to console

File: src/metrics/BusFactor.ts (hunk 1 above)
Comment (tianayjlin) [reply src/metrics/BusFactor.ts:91]: checked in line 47 if response or response.data was empty/null

**Pull request closed without merging**